### ISSUE TITLE
ICE when compiling ICU4X collator in an erroneous state

### ISSUE THREAD

**Issue by hsivonen** [rust-lang/rust#96311]
### Code

1. Clone https://github.com/hsivonen/icu4x
2. Switch to branch `ice_2022-04-22`
3. `cd experimental/collator`
4. `cargo +nightly test` (Note: `+nightly` is required to use nightly, because the project has a toolchain file that specifies 1.60)

### Meta

`rustc --version --verbose`:
```
rustc 1.62.0-nightly (de1bc0008 2022-04-21)
binary: rustc
commit-hash: de1bc0008be096cf7ed67b93402250d3b3e480d0
commit-date: 2022-04-21
host: x86_64-unknown-linux-gnu
release: 1.62.0-nightly
LLVM version: 14.0.1
```

### Additional info

It appears that the ICE was introduced between Rust 1.59 and Rust 1.60. `cargo test`, which uses 1.60 from the toolchain file, also shows the ICE. `RUSTC_BOOTSTRAP=1 RUST_BACKTRACE=1 cargo +1.59 -Z namespaced-features test` does not show the ICE.

### Error output

<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'rustc' panicked at 'assertion failed: !new_self_ty.has_escaping_bound_vars()', compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:1959:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/core/src/panicking.rs:142:14
   2: core::panicking::panic
             at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/core/src/panicking.rs:48:5
   3: <rustc_infer::infer::InferCtxt as rustc_trait_selection::traits::error_reporting::InferCtxtPrivExt>::mk_trait_obligation_with_new_self_ty
   4: <rustc_infer::infer::InferCtxt as rustc_trait_selection::traits::error_reporting::suggestions::InferCtxtExt>::suggest_add_reference_to_arg::{closure#1}
   5: <rustc_infer::infer::InferCtxt as rustc_trait_selection::traits::error_reporting::suggestions::InferCtxtExt>::suggest_add_reference_to_arg
   6: <rustc_infer::infer::InferCtxt as rustc_trait_selection::traits::error_reporting::InferCtxtExt>::report_selection_error
   7: <rustc_infer::infer::InferCtxt as rustc_trait_selection::traits::error_reporting::InferCtxtPrivExt>::report_fulfillment_error
   8: <rustc_infer::infer::InferCtxt as rustc_trait_selection::traits::error_reporting::InferCtxtExt>::report_fulfillment_errors
   9: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_argument_types
  10: <rustc_typeck::check::fn_ctxt::FnCtxt>::confirm_builtin_call
  11: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_call
  12: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_kind
  13: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_with_expectation
  14: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_kind
  15: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_with_expectation
  16: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_stmt
  17: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_block_with_expected
  18: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_kind
  19: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_with_expectation
  20: rustc_typeck::check::check::check_fn
  21: <rustc_infer::infer::InferCtxtBuilder>::enter::<&rustc_middle::ty::context::TypeckResults, <rustc_typeck::check::inherited::InheritedBuilder>::enter<rustc_typeck::check::typeck_with_fallback<rustc_typeck::check::typeck::{closure#0}>::{closure#1}, &rustc_middle::ty::context::TypeckResults>::{closure#0}>
  22: rustc_typeck::check::typeck
  23: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle::ty::context::TyCtxt, rustc_span::def_id::LocalDefId, &rustc_middle::ty::context::TypeckResults>
  24: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<rustc_span::def_id::LocalDefId, &rustc_middle::ty::context::TypeckResults>>
  25: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::typeck
  26: <rustc_middle::hir::map::Map>::par_body_owners::<rustc_typeck::check::typeck_item_bodies::{closure#0}>
  27: rustc_typeck::check::typeck_item_bodies
  28: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle::ty::context::TyCtxt, (), ()>
  29: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<(), ()>>
  30: rustc_query_system::query::plumbing::get_query::<rustc_query_impl::queries::typeck_item_bodies, rustc_query_impl::plumbing::QueryCtxt>
  31: <rustc_session::session::Session>::time::<(), rustc_typeck::check_crate::{closure#7}>
  32: rustc_typeck::check_crate
  33: rustc_interface::passes::analysis
  34: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle::ty::context::TyCtxt, (), core::result::Result<(), rustc_errors::ErrorGuaranteed>>
  35: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<(), core::result::Result<(), rustc_errors::ErrorGuaranteed>>>
  36: rustc_query_system::query::plumbing::get_query::<rustc_query_impl::queries::analysis, rustc_query_impl::plumbing::QueryCtxt>
  37: <rustc_interface::passes::QueryContext>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}::{closure#3}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>
  38: <rustc_interface::interface::Compiler>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}, core::result::Result<core::option::Option<rustc_interface::queries::Linker>, rustc_errors::ErrorGuaranteed>>
  39: rustc_span::with_source_map::<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_interface::interface::create_compiler_and_run<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#1}>
  40: rustc_interface::interface::create_compiler_and_run::<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>
  41: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.62.0-nightly (de1bc0008 2022-04-21) running on x86_64-unknown-linux-gnu

note: compiler flags: -C embed-bitcode=no -C debuginfo=2 -C incremental

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [typeck] type-checking `tests::test_basic`
#1 [typeck_item_bodies] type-checking all item bodies
#2 [analysis] running analysis passes on this crate
end of query stack
error: internal compiler error: no warnings or errors encountered even though `delayed_good_path_bugs` issued

error: internal compiler error: trimmed_def_paths constructed
  |
  = note: delayed at    0: <rustc_errors::Handler>::delay_good_path_bug::<&str>
             1: rustc_middle::ty::print::pretty::trimmed_def_paths
             2: <rustc_middle::dep_graph::dep_node::DepKind as rustc_query_system::dep_graph::DepKind>::with_deps::<<rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task_impl<rustc_middle::ty::context::TyCtxt, (), std::collections::hash::map::HashMap<rustc_span::def_id::DefId, rustc_span::symbol::Symbol, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>>::{closure#0}, std::collections::hash::map::HashMap<rustc_span::def_id::DefId, rustc_span::symbol::Symbol, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>>
             3: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle::ty::context::TyCtxt, (), std::collections::hash::map::HashMap<rustc_span::def_id::DefId, rustc_span::symbol::Symbol, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>>
             4: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::ArenaCache<(), std::collections::hash::map::HashMap<rustc_span::def_id::DefId, rustc_span::symbol::Symbol, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>>>
             5: rustc_query_system::query::plumbing::get_query::<rustc_query_impl::queries::trimmed_def_paths, rustc_query_impl::plumbing::QueryCtxt>
             6: <rustc_middle::ty::print::pretty::FmtPrinter as rustc_middle::ty::print::Printer>::print_def_path
             7: <rustc_middle::ty::print::pretty::FmtPrinter as rustc_middle::ty::print::Printer>::default_print_def_path
             8: <rustc_middle::ty::print::pretty::FmtPrinter as rustc_middle::ty::print::Printer>::print_def_path
             9: <rustc_middle::ty::print::pretty::FmtPrinter as rustc_middle::ty::print::pretty::PrettyPrinter>::pretty_print_type
            10: <rustc_middle::ty::TraitPredicate as rustc_middle::ty::print::Print<rustc_middle::ty::print::pretty::FmtPrinter>>::print
            11: <rustc_middle::ty::print::pretty::FmtPrinter as rustc_middle::ty::print::pretty::PrettyPrinter>::in_binder::<rustc_middle::ty::TraitPredicate>
            12: <rustc_middle::ty::sty::Binder<rustc_middle::ty::TraitPredicate> as core::fmt::Display>::fmt
            13: core::fmt::write
                       at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/core/src/fmt/mod.rs:1194:17
            14: core::fmt::Write::write_fmt
                       at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/core/src/fmt/mod.rs:186:9
            15: alloc::fmt::format
                       at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/alloc/src/fmt.rs:597:5
            16: <rustc_infer::infer::InferCtxt as rustc_trait_selection::traits::error_reporting::InferCtxtExt>::report_selection_error
            17: <rustc_infer::infer::InferCtxt as rustc_trait_selection::traits::error_reporting::InferCtxtPrivExt>::report_fulfillment_error
            18: <rustc_infer::infer::InferCtxt as rustc_trait_selection::traits::error_reporting::InferCtxtExt>::report_fulfillment_errors
            19: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_argument_types
            20: <rustc_typeck::check::fn_ctxt::FnCtxt>::confirm_builtin_call
            21: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_call
            22: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_kind
            23: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_with_expectation
            24: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_kind
            25: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_with_expectation
            26: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_stmt
            27: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_block_with_expected
            28: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_kind
            29: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_with_expectation
            30: rustc_typeck::check::check::check_fn
            31: <rustc_infer::infer::InferCtxtBuilder>::enter::<&rustc_middle::ty::context::TypeckResults, <rustc_typeck::check::inherited::InheritedBuilder>::enter<rustc_typeck::check::typeck_with_fallback<rustc_typeck::check::typeck::{closure#0}>::{closure#1}, &rustc_middle::ty::context::TypeckResults>::{closure#0}>
            32: rustc_typeck::check::typeck
            33: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle::ty::context::TyCtxt, rustc_span::def_id::LocalDefId, &rustc_middle::ty::context::TypeckResults>
            34: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<rustc_span::def_id::LocalDefId, &rustc_middle::ty::context::TypeckResults>>
            35: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::typeck
            36: <rustc_middle::hir::map::Map>::par_body_owners::<rustc_typeck::check::typeck_item_bodies::{closure#0}>
            37: rustc_typeck::check::typeck_item_bodies
            38: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle::ty::context::TyCtxt, (), ()>
            39: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<(), ()>>
            40: rustc_query_system::query::plumbing::get_query::<rustc_query_impl::queries::typeck_item_bodies, rustc_query_impl::plumbing::QueryCtxt>
            41: <rustc_session::session::Session>::time::<(), rustc_typeck::check_crate::{closure#7}>
            42: rustc_typeck::check_crate
            43: rustc_interface::passes::analysis
            44: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle::ty::context::TyCtxt, (), core::result::Result<(), rustc_errors::ErrorGuaranteed>>
            45: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<(), core::result::Result<(), rustc_errors::ErrorGuaranteed>>>
            46: rustc_query_system::query::plumbing::get_query::<rustc_query_impl::queries::analysis, rustc_query_impl::plumbing::QueryCtxt>
            47: <rustc_interface::passes::QueryContext>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}::{closure#3}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>
            48: <rustc_interface::interface::Compiler>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}, core::result::Result<core::option::Option<rustc_interface::queries::Linker>, rustc_errors::ErrorGuaranteed>>
            49: rustc_span::with_source_map::<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_interface::interface::create_compiler_and_run<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#1}>
            50: rustc_interface::interface::create_compiler_and_run::<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>
            51: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>
            52: std::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>
            53: <<std::thread::Builder>::spawn_unchecked_<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
            54: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
                       at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/alloc/src/boxed.rs:1866:9
            55: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
                       at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/alloc/src/boxed.rs:1866:9
            56: std::sys::unix::thread::Thread::new::thread_start
                       at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/std/src/sys/unix/thread.rs:108:17
            57: start_thread
                       at /build/glibc-sMfBJT/glibc-2.31/nptl/pthread_create.c:477:8
            58: clone
                       at /build/glibc-sMfBJT/glibc-2.31/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:95
          

thread 'rustc' panicked at 'Box<dyn Any>', compiler/rustc_errors/src/lib.rs:1347:13
stack backtrace:
   0:     0x7f91d1f49a0d - std::backtrace_rs::backtrace::libunwind::trace::h776940a7aa4a188c
                               at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7f91d1f49a0d - std::backtrace_rs::backtrace::trace_unsynchronized::ha43c91b5ec2ed5c2
                               at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f91d1f49a0d - std::sys_common::backtrace::_print_fmt::h4eeab28a1c89781c
                               at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x7f91d1f49a0d - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hcd8bee1eda901d28
                               at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x7f91d1fa514c - core::fmt::write::h2cf1b656d8dc466b
                               at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/core/src/fmt/mod.rs:1194:17
   5:     0x7f91d1f3b101 - std::io::Write::write_fmt::hcac743f8b5a3865c
                               at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/std/src/io/mod.rs:1655:15
   6:     0x7f91d1f4c725 - std::sys_common::backtrace::_print::h7304aa318dd26d2b
                               at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x7f91d1f4c725 - std::sys_common::backtrace::print::hc8743a1f65280f03
                               at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x7f91d1f4c725 - std::panicking::default_hook::{{closure}}::h7b77f14856e3bc5e
                               at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/std/src/panicking.rs:295:22
   9:     0x7f91d1f4c399 - std::panicking::default_hook::h8570c4038c522cc7
                               at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/std/src/panicking.rs:314:9
  10:     0x7f91d26eabb1 - rustc_driver[a08103eecb06e073]::DEFAULT_HOOK::{closure#0}::{closure#0}
  11:     0x7f91c394f993 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h0d0dcc6c367e1176
                               at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/alloc/src/boxed.rs:1880:9
  12:     0x7f91c39496cd - proc_macro::bridge::client::<impl proc_macro::bridge::Bridge>::enter::{{closure}}::{{closure}}::h2efc5064d15af8da
                               at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/proc_macro/src/bridge/client.rs:335:21
  13:     0x7f91d1f4cef6 - std::panicking::rust_panic_with_hook::ha183336e877ce3b2
                               at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/std/src/panicking.rs:702:17
  14:     0x7f91d3832b41 - std[5b32ac861a1e9abb]::panicking::begin_panic::<rustc_errors[12cdfdb750d72798]::ExplicitBug>::{closure#0}
  15:     0x7f91d38328b6 - std[5b32ac861a1e9abb]::sys_common::backtrace::__rust_end_short_backtrace::<std[5b32ac861a1e9abb]::panicking::begin_panic<rustc_errors[12cdfdb750d72798]::ExplicitBug>::{closure#0}, !>
  16:     0x7f91d383cfdf - std[5b32ac861a1e9abb]::panicking::begin_panic::<rustc_errors[12cdfdb750d72798]::ExplicitBug>
  17:     0x7f91d3839c86 - std[5b32ac861a1e9abb]::panic::panic_any::<rustc_errors[12cdfdb750d72798]::ExplicitBug>
  18:     0x7f91d4fef4a7 - <rustc_errors[12cdfdb750d72798]::HandlerInner as core[c5eed4b94c147a27]::ops::drop::Drop>::drop
  19:     0x7f91d472a248 - core[c5eed4b94c147a27]::ptr::drop_in_place::<rustc_session[564fc83cff9a2edb]::parse::ParseSess>
  20:     0x7f91d472cb43 - <alloc[4da71a5e6ab103ef]::rc::Rc<rustc_session[564fc83cff9a2edb]::session::Session> as core[c5eed4b94c147a27]::ops::drop::Drop>::drop
  21:     0x7f91d47298bd - core[c5eed4b94c147a27]::ptr::drop_in_place::<rustc_interface[d482f63d168e7015]::interface::Compiler>
  22:     0x7f91d472965f - rustc_span[6d64260c30b5c39]::with_source_map::<core[c5eed4b94c147a27]::result::Result<(), rustc_errors[12cdfdb750d72798]::ErrorGuaranteed>, rustc_interface[d482f63d168e7015]::interface::create_compiler_and_run<core[c5eed4b94c147a27]::result::Result<(), rustc_errors[12cdfdb750d72798]::ErrorGuaranteed>, rustc_driver[a08103eecb06e073]::run_compiler::{closure#1}>::{closure#1}>
  23:     0x7f91d4713954 - rustc_interface[d482f63d168e7015]::interface::create_compiler_and_run::<core[c5eed4b94c147a27]::result::Result<(), rustc_errors[12cdfdb750d72798]::ErrorGuaranteed>, rustc_driver[a08103eecb06e073]::run_compiler::{closure#1}>
  24:     0x7f91d46fcb61 - <scoped_tls[7d63459b3a6bde02]::ScopedKey<rustc_span[6d64260c30b5c39]::SessionGlobals>>::set::<rustc_interface[d482f63d168e7015]::interface::run_compiler<core[c5eed4b94c147a27]::result::Result<(), rustc_errors[12cdfdb750d72798]::ErrorGuaranteed>, rustc_driver[a08103eecb06e073]::run_compiler::{closure#1}>::{closure#0}, core[c5eed4b94c147a27]::result::Result<(), rustc_errors[12cdfdb750d72798]::ErrorGuaranteed>>
  25:     0x7f91d471612f - std[5b32ac861a1e9abb]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[d482f63d168e7015]::util::run_in_thread_pool_with_globals<rustc_interface[d482f63d168e7015]::interface::run_compiler<core[c5eed4b94c147a27]::result::Result<(), rustc_errors[12cdfdb750d72798]::ErrorGuaranteed>, rustc_driver[a08103eecb06e073]::run_compiler::{closure#1}>::{closure#0}, core[c5eed4b94c147a27]::result::Result<(), rustc_errors[12cdfdb750d72798]::ErrorGuaranteed>>::{closure#0}, core[c5eed4b94c147a27]::result::Result<(), rustc_errors[12cdfdb750d72798]::ErrorGuaranteed>>
  26:     0x7f91d4716269 - <<std[5b32ac861a1e9abb]::thread::Builder>::spawn_unchecked_<rustc_interface[d482f63d168e7015]::util::run_in_thread_pool_with_globals<rustc_interface[d482f63d168e7015]::interface::run_compiler<core[c5eed4b94c147a27]::result::Result<(), rustc_errors[12cdfdb750d72798]::ErrorGuaranteed>, rustc_driver[a08103eecb06e073]::run_compiler::{closure#1}>::{closure#0}, core[c5eed4b94c147a27]::result::Result<(), rustc_errors[12cdfdb750d72798]::ErrorGuaranteed>>::{closure#0}, core[c5eed4b94c147a27]::result::Result<(), rustc_errors[12cdfdb750d72798]::ErrorGuaranteed>>::{closure#1} as core[c5eed4b94c147a27]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  27:     0x7f91d1f56e43 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h4c4cba3343dc03c1
                               at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/alloc/src/boxed.rs:1866:9
  28:     0x7f91d1f56e43 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h2bb45661b1290f4d
                               at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/alloc/src/boxed.rs:1866:9
  29:     0x7f91d1f56e43 - std::sys::unix::thread::Thread::new::thread_start::h3d8e7bbd329b246a
                               at /rustc/de1bc0008be096cf7ed67b93402250d3b3e480d0/library/std/src/sys/unix/thread.rs:108:17
  30:     0x7f91d1e6b609 - start_thread
                               at /build/glibc-sMfBJT/glibc-2.31/nptl/pthread_create.c:477:8
  31:     0x7f91d1d84163 - clone
                               at /build/glibc-sMfBJT/glibc-2.31/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:95
  32:                0x0 - <unknown>

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.62.0-nightly (de1bc0008 2022-04-21) running on x86_64-unknown-linux-gnu

note: compiler flags: -C embed-bitcode=no -C debuginfo=2 -C incremental

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
thread panicked while panicking. aborting.
error: could not compile `icu_collator`

Caused by:
  process didn't exit successfully: `rustc --crate-name icu_collator --edition=2018 experimental/collator/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --emit=dep-info,link -C embed-bitcode=no -C debuginfo=2 --test --cfg 'feature="default"' -C metadata=3fc0b0d4367a31df -C extra-filename=-3fc0b0d4367a31df --out-dir /opt/Projects/icu4x/target/debug/deps -C incremental=/opt/Projects/icu4x/target/debug/incremental -L dependency=/opt/Projects/icu4x/target/debug/deps --extern arraystring=/opt/Projects/icu4x/target/debug/deps/libarraystring-6a8fd1c61113bf09.rlib --extern atoi=/opt/Projects/icu4x/target/debug/deps/libatoi-574e73dc98b9621c.rlib --extern displaydoc=/opt/Projects/icu4x/target/debug/deps/libdisplaydoc-e9f1bbbdbf69c3b6.so --extern icu_char16trie=/opt/Projects/icu4x/target/debug/deps/libicu_char16trie-de5e7fbd95d5855e.rlib --extern icu_codepointtrie=/opt/Projects/icu4x/target/debug/deps/libicu_codepointtrie-56530ffc0d2d6560.rlib --extern icu_locid=/opt/Projects/icu4x/target/debug/deps/libicu_locid-331b544e7e9136e1.rlib --extern icu_normalizer=/opt/Projects/icu4x/target/debug/deps/libicu_normalizer-49232aef8af6f5ef.rlib --extern icu_properties=/opt/Projects/icu4x/target/debug/deps/libicu_properties-0359c6ad5e1c5d98.rlib --extern icu_provider=/opt/Projects/icu4x/target/debug/deps/libicu_provider-da8e2a821666f767.rlib --extern icu_testdata=/opt/Projects/icu4x/target/debug/deps/libicu_testdata-d2aabfc5dcfb7a67.rlib --extern postcard=/opt/Projects/icu4x/target/debug/deps/libpostcard-a3ffbc6de844bde3.rlib --extern smallvec=/opt/Projects/icu4x/target/debug/deps/libsmallvec-21d1ce955946ca7f.rlib --extern utf8_iter=/opt/Projects/icu4x/target/debug/deps/libutf8_iter-c9fe58b74f413b57.rlib --extern zerovec=/opt/Projects/icu4x/target/debug/deps/libzerovec-1cf7f5463998d936.rlib` (signal: 6, SIGABRT: process abort signal)```

</p>
</details>

**Comment by hsivonen** [comment]
After using Rust 1.59 to fix the errors in the collator code, Rust 1.60 is able to compile it without ICE.

**Comment by Enselic** [comment]
Triage: Can this still be reproduced? If yes, it would be great if you could minimize the reproduction.
